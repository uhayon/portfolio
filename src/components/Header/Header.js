import React from 'react';
import PropTypes from 'prop-types';

import NavigationBar from './NavigationBar/NavigationBar';
import Logo from '../Logo/Logo';
import ToggleNavigationBarButton from './ToggleNavigationBarButton/ToggleNavigationBarButton';

import navigationButtons from './NavigationBarButtons.json';

import styles from './Header.module.scss';

class Header extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isMobileScreen: this.isMobileScreen(),
      navigationBarExpanded: false,
      navigationBarFixed: this.getNavigationBarFixed()
    };
  }

  componentDidMount() {
    window.addEventListener('resize', this.setIsMobileScreen);
    window.addEventListener('scroll', this.setNavigationBarFixed);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.setIsMobileScreen);
    window.removeEventListener('scroll', this.setNavigationBarFixed);
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      nextProps.selectedLanguage !== this.props.selectedLanguage || 
      nextState.navigationBarExpanded !== this.state.navigationBarExpanded ||
      nextState.isMobileScreen !== this.state.isMobileScreen ||
      nextState.navigationBarFixed !== this.state.navigationBarFixed
    );
  }

  getNavigationBarFixed = () => {
    return window.scrollY >= 30;
  }

  setNavigationBarFixed = () => {
    this.setState(prevState => ({
      ...prevState,
      navigationBarFixed: this.getNavigationBarFixed()
    }));
  }

  setIsMobileScreen = () => {
    const isMobileScreen = this.isMobileScreen();
    this.setState(prevState => ({
      ...prevState,
      isMobileScreen,
      navigationBarExpanded: isMobileScreen ? prevState.navigationBarExpanded : false
    }));
  }

  isMobileScreen = () => {
    return window.innerWidth < 990;
  }

  handleToggleNavigationBarExpansion = () => {
    this.setState(prevState => ({
      ...prevState,
      navigationBarExpanded: !prevState.navigationBarExpanded
    }), () => {
      const body = document.querySelector('body');
      this.state.navigationBarExpanded ? body.classList.add('overflow-hidden') : body.classList.remove('overflow-hidden');
    });
  }

  render() {
    const { selectedLanguage, handleLanguageSwitch } = this.props;
    const { navigationBarExpanded, isMobileScreen, navigationBarFixed} = this.state;
    
    return (
      <header className={`${styles.header} ${navigationBarFixed ? styles.fixed : ''}`}>
        <a href='#home' className={styles.linkLogo}>
          <Logo />
        </a>
        <ToggleNavigationBarButton
          isExpandButton
          isMobileScreen={isMobileScreen}
          handleNavButtonPress={this.handleToggleNavigationBarExpansion} />
        <NavigationBar 
          selectedLanguage={selectedLanguage} 
          toggleSidebarExpansion={this.handleToggleNavigationBarExpansion}
          handleLanguageSwitch={handleLanguageSwitch}
          isExpanded={navigationBarExpanded} 
          isSideBar={isMobileScreen}
          buttons={navigationButtons} />
      </header>
    );
  }
}

Header.propTypes = {
  selectedLanguage: PropTypes.string.isRequired,
  handleLanguageSwitch: PropTypes.func.isRequired
}

export default Header;