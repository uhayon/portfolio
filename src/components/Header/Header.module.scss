@import '../../globalStyles/colors.scss';

.header {
    background-color: $black;
    height: 50px;
    z-index: 1000;
    padding: .5rem 1rem;
    display: flex;
    justify-content: space-between;
    align-items: center;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    transition: height 0.1s linear;

    &.fixed {
        height: 40px;
        .linkLogo {
            height: 25px;
        }

        li > a {
            font-size: 1.1rem;
        }
    }

    .linkLogo {
        height: 30px;
        transition: height 0.2s ease-out;
    }
}