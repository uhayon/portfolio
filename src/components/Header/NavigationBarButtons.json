[
    {
        "title": {
            "EN": "Skills",
            "ES": "Habilidades"
        },
        "sectionId": "section-skills"
    },
    {
        "title": {
            "EN": "Projects",
            "ES": "Proyectos"
        },
        "sectionId": "section-projects"
    },
    {
        "title": {
            "EN": "Contact Me",
            "ES": "Contactame"
        },
        "sectionId": "section-contact"
    }
]