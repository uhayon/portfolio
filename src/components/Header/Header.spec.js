import React from 'react';
import { shallow, mount } from 'enzyme';
import Header from './Header';

describe('Mounting Header', () => {
  let component;
  const handleLanguageSwitch = jest.fn(() => 'ES');
  beforeEach(() => component = <Header handleLanguageSwitch={handleLanguageSwitch} selectedLanguage='EN' />);

  it('Must match snapshot without crashing', () => {
    const wrapper = shallow(component);
    expect(wrapper).toMatchSnapshot();
  });

  it('Must have selectedLanguage and setSelectedLanguage props', () => {
    const wrapper = mount(component);

    expect(wrapper.prop('selectedLanguage')).toBeDefined();
    expect(wrapper.prop('handleLanguageSwitch')).toBeDefined();
    expect(typeof wrapper.prop('handleLanguageSwitch')).toEqual('function');

    wrapper.unmount()
  });
});