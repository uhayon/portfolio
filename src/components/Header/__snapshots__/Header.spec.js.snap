// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mounting Header Must match snapshot without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Header
    handleLanguageSwitch={[MockFunction]}
    selectedLanguage="EN"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <a
          className="linkLogo"
          href="#home"
        >
          <Logo />
        </a>,
        <ToggleNavigationBarButton
          handleNavButtonPress={[Function]}
          isExpandButton={true}
          isMobileScreen={false}
        />,
        <NavigationBar
          buttons={
            Array [
              Object {
                "sectionId": "skills",
                "title": Object {
                  "EN": "Skills",
                  "ES": "Habilidades",
                },
              },
              Object {
                "sectionId": "projects",
                "title": Object {
                  "EN": "Projects",
                  "ES": "Proyectos",
                },
              },
              Object {
                "sectionId": "contact",
                "title": Object {
                  "EN": "Contact Me",
                  "ES": "Contactame",
                },
              },
            ]
          }
          handleLanguageSwitch={[MockFunction]}
          isExpanded={false}
          isSideBar={false}
          selectedLanguage="EN"
          toggleSidebarExpansion={[Function]}
        />,
      ],
      "className": "header ",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Logo />,
          "className": "linkLogo",
          "href": "#home",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "handleNavButtonPress": [Function],
          "isExpandButton": true,
          "isMobileScreen": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "buttons": Array [
            Object {
              "sectionId": "skills",
              "title": Object {
                "EN": "Skills",
                "ES": "Habilidades",
              },
            },
            Object {
              "sectionId": "projects",
              "title": Object {
                "EN": "Projects",
                "ES": "Proyectos",
              },
            },
            Object {
              "sectionId": "contact",
              "title": Object {
                "EN": "Contact Me",
                "ES": "Contactame",
              },
            },
          ],
          "handleLanguageSwitch": [MockFunction],
          "isExpanded": false,
          "isSideBar": false,
          "selectedLanguage": "EN",
          "toggleSidebarExpansion": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "header",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <a
            className="linkLogo"
            href="#home"
          >
            <Logo />
          </a>,
          <ToggleNavigationBarButton
            handleNavButtonPress={[Function]}
            isExpandButton={true}
            isMobileScreen={false}
          />,
          <NavigationBar
            buttons={
              Array [
                Object {
                  "sectionId": "skills",
                  "title": Object {
                    "EN": "Skills",
                    "ES": "Habilidades",
                  },
                },
                Object {
                  "sectionId": "projects",
                  "title": Object {
                    "EN": "Projects",
                    "ES": "Proyectos",
                  },
                },
                Object {
                  "sectionId": "contact",
                  "title": Object {
                    "EN": "Contact Me",
                    "ES": "Contactame",
                  },
                },
              ]
            }
            handleLanguageSwitch={[MockFunction]}
            isExpanded={false}
            isSideBar={false}
            selectedLanguage="EN"
            toggleSidebarExpansion={[Function]}
          />,
        ],
        "className": "header ",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Logo />,
            "className": "linkLogo",
            "href": "#home",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "handleNavButtonPress": [Function],
            "isExpandButton": true,
            "isMobileScreen": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "buttons": Array [
              Object {
                "sectionId": "skills",
                "title": Object {
                  "EN": "Skills",
                  "ES": "Habilidades",
                },
              },
              Object {
                "sectionId": "projects",
                "title": Object {
                  "EN": "Projects",
                  "ES": "Proyectos",
                },
              },
              Object {
                "sectionId": "contact",
                "title": Object {
                  "EN": "Contact Me",
                  "ES": "Contactame",
                },
              },
            ],
            "handleLanguageSwitch": [MockFunction],
            "isExpanded": false,
            "isSideBar": false,
            "selectedLanguage": "EN",
            "toggleSidebarExpansion": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "header",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
